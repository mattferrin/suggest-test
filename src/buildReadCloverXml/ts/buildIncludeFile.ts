import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
