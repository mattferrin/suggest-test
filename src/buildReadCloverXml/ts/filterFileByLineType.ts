import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
