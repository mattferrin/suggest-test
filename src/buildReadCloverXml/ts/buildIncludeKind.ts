import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    const isStatementKeeper =
      (file?.statements ?? 0) > 1 && (file?.uncoveredStatements ?? 0) > 0;
    const isConditionalKeeper =
      (file?.conditionals ?? 0) > 1 && (file?.uncoveredConditionals ?? 0) > 0;

    if (!input.flags.includeConditionals) {
      return isStatementKeeper;
    } else if (!input.flags.includeStatements) {
      return isConditionalKeeper;
    } else {
      return isStatementKeeper || isConditionalKeeper;
    }
  };
}
