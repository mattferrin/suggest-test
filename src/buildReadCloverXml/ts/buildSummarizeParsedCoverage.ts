import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
