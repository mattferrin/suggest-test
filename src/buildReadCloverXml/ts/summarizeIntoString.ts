/* eslint-disable functional/functional-parameters */
/* eslint-disable @typescript-eslint/restrict-template-expressions */

import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function summarizeIntoString(
  input: MeowOutput,
  summaryArray: ReadonlyArray<SummaryItem | undefined>
): string {
  const entity = [
    ...(input.flags.includeConditionals ? ["conditionals"] : []),
    ...(input.flags.includeStatements ? ["statements"] : []),
  ];

  if (summaryArray.length === 0) {
    return `\nNo uncovered ${entity.join(" or ")} were found.\n`;
  } else {
    const outputLines = summaryArray.map((line) => {
      return `${line?.path ?? ""} ${new Intl.NumberFormat("en-US", {
        maximumSignificantDigits: 2,
      }).format(line?.score ?? 0)}`;
    });

    return [
      `\nFiles suggested to increase coverage of ${entity.join(" and ")}:\n`,
      ...outputLines,
      "",
    ]
      .filter((line) => line !== null)
      .join("\n");
  }
}
