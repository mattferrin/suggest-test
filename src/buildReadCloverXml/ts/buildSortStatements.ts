import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
