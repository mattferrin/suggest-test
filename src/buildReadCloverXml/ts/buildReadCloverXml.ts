import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
