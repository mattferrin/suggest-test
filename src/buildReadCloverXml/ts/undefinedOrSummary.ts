import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export interface SummaryItem {
  readonly conditionals: number | null | undefined;
  readonly path: string | null | undefined;
  readonly statements: number | null | undefined;
  readonly uncoveredConditionals: number;
  readonly uncoveredStatements: number;
  readonly score: number;
  readonly hasUnmarkedConditional: boolean;
}

export function undefinedOrSummary(
  summary: ReadonlyArray<SummaryItem | undefined> | undefined
): ReadCloverXmlOutput {
  if (summary === undefined || summary === null) {
    return {
      tag: "errors",
      errors: [
        new Error(
          "The parsed clover.xml has an undefined or null property somewhere"
        ),
      ],
    };
  } else {
    return {
      tag: "summary",
      summaryArray: summary,
      summaryString: { tag: "not-ready" },
    };
  }
}
