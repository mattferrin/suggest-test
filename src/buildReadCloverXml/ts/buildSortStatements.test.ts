import { MeowOutput } from "../../bin/ts/cli";
import * as buildIncludeFile from "./buildIncludeFile";
import * as buildIncludeKind from "./buildIncludeKind";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { buildSortStatements } from "./buildSortStatements";
import * as summarizeIntoString from "./summarizeIntoString";
import { SummaryItem } from "./undefinedOrSummary";

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: Parameters<typeof buildSortStatements>[0];
  output: Parameters<ReturnType<typeof buildSortStatements>>[0];
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */
    const summarizeIntoStringSpy = jest
      .spyOn(summarizeIntoString, "summarizeIntoString")
      .mockImplementation(() => "summarizeIntoString result");
    const buildIncludeKindSpy = jest
      .spyOn(buildIncludeKind, "buildIncludeKind")
      .mockImplementation(() => () => true);
    const buildIncludeFileSpy = jest
      .spyOn(buildIncludeFile, "buildIncludeFile")
      .mockImplementation(() => () => true);

    /** unit under test */
    const sortStatements = buildSortStatements(input.input);
    const result = sortStatements(input.output);

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
        summarizeIntoStringCalls: summarizeIntoStringSpy.mock.calls,
        buildIncludeKindCalls: buildIncludeKindSpy.mock.calls,
        buildIncludeFileCalls: buildIncludeFileSpy.mock.calls,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "3d99af10-0b75-4cb9-a3b0-2358555af621",
  comment: "returns errors result with error item",
  input: "not relevant" as any,
  output: { tag: "errors", errors: [new Error("expect this errors item")] },
});
snap({
  stabileSortId: "d117a7b1-b287-4943-906b-2cc8f1369786",
  comment: "forwards empty `summaryArray`",
  input: { flags: { maxItems: 1 } } as MeowOutput,
  output: {
    tag: "summary",
    summaryArray: [],
    summaryString: "not relevant",
  } as ReadCloverXmlOutput,
});
snap({
  stabileSortId: "675ec71a-2b9b-49f8-89d4-44a72839266b",
  comment: "sorts score 2 into top position",
  input: { flags: { maxItems: 1 } } as MeowOutput,
  output: {
    tag: "summary",
    summaryArray: [{ score: 1 }, { score: 2 }] as SummaryItem[],
    summaryString: "not relevant",
  } as ReadCloverXmlOutput,
});
snap({
  stabileSortId: "41c5f18b-0f55-406e-8e8f-9048a6b442d4",
  comment: "does NOT error when files undefined",
  input: { flags: { maxItems: 2 } } as MeowOutput,
  output: {
    tag: "summary",
    summaryArray: [undefined, undefined, undefined] as undefined[],
    summaryString: "not relevant",
  } as ReadCloverXmlOutput,
});
snap({
  stabileSortId: "3b12967f-716b-4913-abfb-ff07fbaf9428",
  comment: "does NOT error when score undefined",
  input: { flags: { maxItems: 3 } } as MeowOutput,
  output: {
    tag: "summary",
    summaryArray: [{}, {}] as SummaryItem[],
    summaryString: "not relevant",
  } as ReadCloverXmlOutput,
});
