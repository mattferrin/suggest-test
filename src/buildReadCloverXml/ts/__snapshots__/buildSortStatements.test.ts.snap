// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3b12967f-716b-4913-abfb-ff07fbaf9428 1`] = `
{
  "does NOT error when score undefined": {
    "buildIncludeFileCalls": [
      [
        {
          "flags": {
            "maxItems": 3,
          },
        },
      ],
    ],
    "buildIncludeKindCalls": [
      [
        {
          "flags": {
            "maxItems": 3,
          },
        },
      ],
    ],
    "result": {
      "summaryArray": [
        {},
        {},
      ],
      "summaryString": "summarizeIntoString result",
      "tag": "summary",
    },
    "summarizeIntoStringCalls": [
      [
        {
          "flags": {
            "maxItems": 3,
          },
        },
        [
          {},
          {},
        ],
      ],
    ],
  },
}
`;

exports[`3d99af10-0b75-4cb9-a3b0-2358555af621 1`] = `
{
  "returns errors result with error item": {
    "buildIncludeFileCalls": [],
    "buildIncludeKindCalls": [],
    "result": {
      "errors": [
        [Error: expect this errors item],
      ],
      "tag": "errors",
    },
    "summarizeIntoStringCalls": [],
  },
}
`;

exports[`41c5f18b-0f55-406e-8e8f-9048a6b442d4 1`] = `
{
  "does NOT error when files undefined": {
    "buildIncludeFileCalls": [
      [
        {
          "flags": {
            "maxItems": 2,
          },
        },
      ],
    ],
    "buildIncludeKindCalls": [
      [
        {
          "flags": {
            "maxItems": 2,
          },
        },
      ],
    ],
    "result": {
      "summaryArray": [
        undefined,
        undefined,
      ],
      "summaryString": "summarizeIntoString result",
      "tag": "summary",
    },
    "summarizeIntoStringCalls": [
      [
        {
          "flags": {
            "maxItems": 2,
          },
        },
        [
          undefined,
          undefined,
        ],
      ],
    ],
  },
}
`;

exports[`675ec71a-2b9b-49f8-89d4-44a72839266b 1`] = `
{
  "sorts score 2 into top position": {
    "buildIncludeFileCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
      ],
    ],
    "buildIncludeKindCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
      ],
    ],
    "result": {
      "summaryArray": [
        {
          "score": 2,
        },
      ],
      "summaryString": "summarizeIntoString result",
      "tag": "summary",
    },
    "summarizeIntoStringCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
        [
          {
            "score": 2,
          },
        ],
      ],
    ],
  },
}
`;

exports[`d117a7b1-b287-4943-906b-2cc8f1369786 1`] = `
{
  "forwards empty \`summaryArray\`": {
    "buildIncludeFileCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
      ],
    ],
    "buildIncludeKindCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
      ],
    ],
    "result": {
      "summaryArray": [],
      "summaryString": "summarizeIntoString result",
      "tag": "summary",
    },
    "summarizeIntoStringCalls": [
      [
        {
          "flags": {
            "maxItems": 1,
          },
        },
        [],
      ],
    ],
  },
}
`;
